### PR TITLE
Revscriptsys: Weapon absorb percent all and by an element

### PR DESCRIPTION
**Pull request by souzajunior**
**Function**
weapon:absorbPercentAll(value)
weapon:absorbPercent(type, value)

**Example**
weapon:absorbPercentAll(15)
weapon:absorbPercent("fire", 10)

### CODE REVIEW

**Comment by eduardodantas95** [comment]
Tested. 
All ok

File: src/tools.cpp [hunk 1]
```diff
@@ -725,6 +725,16 @@ std::string getCombatName(CombatType_t combatType)
 	return "unknown";
 }
 
+CombatType_t getCombatTypeByName(std::string name)
```
Comment (djarek) [inline src/tools.cpp]:
```suggestion
CombatType_t getCombatTypeByName(const std::string& name)
```

File: src/tools.cpp [hunk 2]
```diff
@@ -725,6 +725,16 @@ std::string getCombatName(CombatType_t combatType)
...
+{
+	for (auto it = combatTypeNames.begin(); it != combatTypeNames.end(); ++it) {
```
Comment (djarek) [inline src/tools.cpp]: Use `find_if`, makes it clearer you just want to find an element in the map:
```
auto it = std::find_if(combatTypeNames.begin(), combatTypeNames.end(),
  [&](std::pair<std::string, CombatType_t> const& pair)
  {
    return pair.second == name;
  });

if (it == combatTypeNames.end())
  return COMBAT_NONE;
else
  return it->first;
```

Alternatively, there should be a reverse-lookup map.

Comment (souzajunior) [reply src/tools.cpp]: Great modification, ty. Just a correction:
`[&](std::pair<std::string, CombatType_t> const& pair)` 
modify to:
`[&](std::pair<CombatType_t, std::string> const& pair)`

File: src/tools.cpp [hunk 3]
```diff
@@ -725,6 +725,20 @@ std::string getCombatName(CombatType_t combatType)
 	return "unknown";
 }
 
+CombatType_t getCombatTypeByName(const std::string& name)
+{
+	auto it = std::find_if(combatTypeNames.begin(), combatTypeNames.end(),
+		[&](std::pair<CombatType_t, std::string> const& pair)
+	{
```
Comment (djarek) [inline src/tools.cpp]: One minor thing: please check if this is consistent with lambda formatting elsewhere. IIRC braces in lambda expressions should be attached to the first line of the expression.

Comment (souzajunior) [reply src/tools.cpp]: It's working fine, I didn't see any problem with this modification. I update with your review, compiled and yeah, all right.

**Comment by EvilHero90** [comment]
This entire thing was actually planned to be in Items/ItemType once it gets a complete workover with item virtualization and such, so maybe instead of having it strictly for weapons only you could adapt it to ItemType for now, so we don't have it double once ItemType gets a re work and is already included.

File: src/tools.cpp (hunk 3 above)
Comment (djarek) [reply src/tools.cpp]: I was referring to putting `{` in the previous line, because that is, I think how it's formatted elsewhere in the project.

**Comment by souzajunior** [comment]
> This entire thing was actually planned to be in Items/ItemType once it gets a complete workover with item virtualization and such, so maybe instead of having it strictly for weapons only you could adapt it to ItemType for now, so we don't have it double once ItemType gets a re work and is already included.

Oh, okay, no problem. My friend told me it was not implemented, so I thought it was better to do a PR here first. Anyway, thank you for reply.

File: src/tools.cpp (hunk 3 above)
Comment (souzajunior) [reply src/tools.cpp]: Oooh, sorry, haha. I guess so, I'm still adapting to the project, Ty.

**Comment by EvilHero90** [comment]
You don't need to close it, i just meant if you moved the code to itemtype instead it would perfectly fit for later on, so we don't have to do it twice :D

**Pull request closed without merging**